<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="four_wheel_steering_robot">

  <xacro:property name="wheel_base" value="2.8"/>
  <xacro:property name="wheel_track" value="1.385"/>
  <xacro:property name="wheel_radius" value="0.3"/>
  <xacro:property name="max_steering_angle" value="0.31"/>
  <xacro:property name="min_steering_angle" value="-0.31"/>
  <xacro:property name="base_mass" value="1703.239"/> <!-- Mass value is incorrect as the intertias are wrong-->
  <xacro:property name="tyre_x" value="1.05"/>
  <xacro:property name="tyre_front_x" value="1.40"/>
  <xacro:property name="tyre_back_x" value="-1.40"/>
  <xacro:property name="tyre_y" value="0.6925"/>
  <xacro:property name="tyre_r" value="0"/>
  <xacro:property name="PI" value="3.1415926835897931"/>
  <!--Below is the thickness of the car's tyre-->
  <xacro:property name="tyre_length" value="0.18"/>

  <xacro:property name="tyre_radius" value="0.33"/>
  <xacro:property name="tyre_mass" value="20"/>

  <xacro:property name="tyre_mass_ixx" value="${((1/2)*tyre_mass*tyre_radius*tyre_radius)}"/>
  <xacro:property name="tyre_mass_iyy" value="${((1/12)*tyre_mass*tyre_length*tyre_length) + ((1/4)*tyre_mass*tyre_radius*tyre_radius)}"/>
  <xacro:property name="tyre_mass_izz" value="${((1/12)*tyre_mass*tyre_length*tyre_length) + ((1/4)*tyre_mass*tyre_radius*tyre_radius)}"/>
  <xacro:property name="tyre_mass_ixz" value="0.0"/>
  <xacro:property name="tyre_mass_ixy" value="0.0"/>
  <xacro:property name="tyre_mass_iyz" value="0.0"/>

  <xacro:property name="str_mass" value="50.00"/>
  <xacro:property name="str_length" value="0.05"/>
  <xacro:property name="str_radius" value="0.1"/>
  <xacro:property name="front_right_offset" value="0.05"/>
  <xacro:property name="str_mass_ixx" value="${((1/2)*str_mass*str_radius*str_radius)}"/>
  <xacro:property name="str_mass_iyy" value="${((1/12)*str_mass*str_length*str_length) + ((1/4)*str_mass*str_radius*str_radius)}"/>
  <xacro:property name="str_mass_izz" value="${((1/12)*str_mass*str_length*str_length) + ((1/4)*str_mass*str_radius*str_radius)}"/>
  
  <xacro:property name="str_angle" value="0.31"/>
  <xacro:property name="car_width" value="1.385"/>
  <xacro:property name="car_width_offset" value="0.05"/>

<xacro:property name="tyre_origin">
        <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
    </xacro:property>
    <xacro:property name="tyre_origin_2">
        <origin xyz="0 0 0" rpy="0 0 ${PI/2}"/>
    </xacro:property>
    <xacro:property name="tyre_origin_3">
        <origin xyz="0 0 0" rpy="0 0 ${-PI/2}"/>
    </xacro:property>

  <material name="white">
      <color rgba="1 1 1 1" />
  </material>

  <material name="orange">
      <color rgba="1 0.3 0.1 1"/>
  </material>

  <material name="blue">
      <color rgba="0.2 0.2 1 1"/>
  </material>

  <material name="black">
      <color rgba="0 0 0 1"/>
  </material>

  <material name="red">
      <color rgba="1 0 0 1"/>
  </material>

  <!-- Define robot links -->
  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 1.57079633"/>
      <geometry>
        <mesh filename="package://4ws_robot/meshes/gen0_colored.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="inertial_joint" type="fixed">
      <parent link="base_link"/>
      <child link="main_mass"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>


  <link name="main_mass" type="fixed">
      <parent link="inertial_joint"/>
      <inertial>
          <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/> <!-- This is the center of mass of Car Body-->
          <mass value="${base_mass}"/>
          <inertia
            ixx="347.195805" ixy="-11.4914985" ixz="18.5070628"
            iyy="2330.10026" iyz="3.97814264"
            izz="2529.41827"/>
      </inertial>
  </link>



    <joint name="back_right_steering_joint" type="revolute">
        <parent link="base_link"/>
        <child link="back_right_steering_link"/>
        <origin xyz="${tyre_back_x} ${-(car_width/2 + 2*str_length)} ${tyre_radius}" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="10000" lower="${-str_angle}" upper="${str_angle}" velocity="1000"/>
    </joint>

    <link name="back_right_steering_link">
        <collision>
            <xacro:insert_block name="tyre_origin"/>
	        <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
       		</geometry>
        </collision>
        <visual>
            <xacro:insert_block name="tyre_origin"/>
            <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
            </geometry>
        </visual>
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${str_mass}"/>
            <inertia
            ixx="${str_mass_ixx}"  ixy="0.000000"  ixz="0.000000"
            iyy="${str_mass_iyy}"  iyz="0.000000"
            izz="${str_mass_izz}"/>
        </inertial>
    </link>


    <joint name="back_right_wheel_joint" type="continuous">
        <parent link="back_right_steering_link"/>
        <child link="back_right_wheel_link"/>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.01"/>
        <limit effort="10000" velocity="5000" />
        <joint_properties damping="10" friction="1.0" />
    </joint>

    <link name="back_right_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <!--mesh filename="package://catvehicle/meshes/ford_escape_wheel.dae"/-->
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>10000000</mu>
                        <mu2>7000</mu2>
                    </ode>
                </friction>
                <bounce>
                    <restitution_coefficient>0.0</restitution_coefficient>
                </bounce>
                <contact>
                    <ode>
                        <soft_cfm>0.001</soft_cfm>
                        <soft_erp>0.5</soft_erp>
                        <kp>10000000</kp>
                        <kd>1</kd>
                    </ode>
                </contact>
            </surface>
        </collision>

        <visual>
            <!--origin xyz="0 0 0" rpy="${PI/2} 0 0"/-->
            <origin xyz="0 ${car_width_offset*2} 0.0" rpy="0 0 ${-PI/2}"/>
            <geometry>
                <!--cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <scale xyz="1 1 1"/-->
                <mesh scale="1.2 1 1" filename="package://4ws_robot/meshes/ezmile_gen0_wheel.obj"/>
            </geometry>
        </visual>
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${tyre_mass}"/>
            <inertia
            ixx="${tyre_mass_ixx}"  ixy="${tyre_mass_ixy}"  ixz="${tyre_mass_ixz}"
            iyy="${tyre_mass_iyy}"  iyz="${tyre_mass_iyz}"
            izz="${tyre_mass_izz}"/>
        </inertial>
    </link>

    <!--Back Left Wheel -->

    <joint name="back_left_steering_joint" type="revolute">
        <parent link="base_link"/>
        <child link="back_left_steering_link"/>
        <origin xyz="${tyre_back_x} ${car_width/2 + 2*str_length} ${tyre_radius}" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="10000" lower="${-str_angle}" upper="${str_angle}" velocity="1000"/>
    </joint>

    <link name="back_left_steering_link">
        <collision>
            <xacro:insert_block name="tyre_origin"/>
	        <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
       		</geometry>
        </collision>
        <visual>
            <xacro:insert_block name="tyre_origin"/>
            <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
            </geometry>
        </visual>
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${str_mass}"/>
            <inertia
            ixx="${str_mass_ixx}"  ixy="0.000000"  ixz="0.000000"
            iyy="${str_mass_iyy}"  iyz="0.000000"
            izz="${str_mass_izz}"/>
        </inertial>
    </link>


    <joint name="back_left_wheel_joint" type="continuous">
        <parent link="back_left_steering_link"/>
        <child link="back_left_wheel_link"/>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.01"/>
        <limit effort="10000" velocity="5000" />
        <joint_properties damping="10" friction="1.0" />
    </joint>
    
    <link name="back_left_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <!--mesh filename="package://ezmile_gen0/meshes/ford_escape_wheel.dae"/-->
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>10000000</mu>
                        <mu2>7000</mu2>
                    </ode>
                </friction>
                <bounce>
                    <restitution_coefficient>0.0</restitution_coefficient>
                </bounce>
                <contact>
                    <ode>
                        <soft_cfm>0.001</soft_cfm>
                        <soft_erp>0.5</soft_erp>
                        <kp>10000000</kp>
                        <kd>1</kd>
                    </ode>
                </contact>
            </surface>
        </collision>
        
        <visual>
            <!--origin xyz="0 0 0" rpy="${PI/2} 0 0"/-->
            <origin xyz="0 ${-car_width_offset*2} 0.0" rpy="0 0 ${PI/2}"/>
            <geometry>
                <!--cylinder length="${tyre_length}" radius="${tyre_radius}"/-->
                <mesh scale="1.2 1 1" filename="package://4ws_robot/meshes/ezmile_gen0_wheel.obj"/>
            </geometry>
        </visual>
        
        <inertial>
        
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${tyre_mass}"/>
            <inertia
            ixx="${tyre_mass_ixx}"  ixy="${tyre_mass_ixy}"  ixz="${tyre_mass_ixz}"
            iyy="${tyre_mass_iyy}"  iyz="${tyre_mass_iyz}"
            izz="${tyre_mass_izz}"/>
        </inertial>
    </link>



    <!--Right Steering-->

    <joint name="front_right_steering_joint" type="revolute">
        <parent link="base_link"/>
        <child link="front_right_steering_link"/>
        <origin xyz="${tyre_front_x} ${-car_width/2 - 2*str_length} ${tyre_radius}" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="10000" lower="${-str_angle}" upper="${str_angle}" velocity="1000"/>
    </joint>

    <link name="front_right_steering_link">
        <collision>
            <xacro:insert_block name="tyre_origin"/>
	        <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
       		</geometry>
        </collision>
        <visual>
            <xacro:insert_block name="tyre_origin"/>
            <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
            </geometry>
        </visual>
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${str_mass}"/>
            <inertia
            ixx="${str_mass_ixx}"  ixy="0.000000"  ixz="0.000000"
            iyy="${str_mass_iyy}"  iyz="0.000000"
            izz="${str_mass_izz}"/>
        </inertial>
    </link>

    <!--Front Right Wheel-->
    <joint name="front_right_wheel_joint" type="continuous">
        <parent link="front_right_steering_link"/>
        <child link="front_right_wheel_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.01"/>
        <limit effort="10000" velocity="5000"/>
        <joint_properties damping="10" friction="1.0" />
    </joint>
    
    <link name="front_right_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <!--origin xyz="0 0 0" rpy="0 0 ${-PI/2}"/-->
            <geometry>
                <cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <!--mesh filename="package://ezmile_gen0/meshes/ford_escape_wheel.dae"/-->
            </geometry>
            
            <surface>
                <friction>
                    <ode>
                        <mu>10000000</mu>
                        <mu2>7000</mu2>
                    </ode>
                </friction>
                <bounce>
                    <restitution_coefficient>0.0</restitution_coefficient>
                </bounce>
                <contact>
                    <ode>
                        <soft_cfm>0.001</soft_cfm>
                        <soft_erp>0.5</soft_erp>
                        <kp>10000000</kp>
                        <kd>1</kd>
                    </ode>
                </contact>
            </surface>
        </collision>


        <visual>
            <!--origin xyz="0 0 0" rpy="${PI/2} 0 0"/-->
            <origin xyz="0 ${car_width_offset*2} 0.0" rpy="0 0 ${-PI/2}"/>
            <geometry>
                <!--cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <scale xyz="1 1 1"/-->
                <mesh scale="1.2 1 1" filename="package://4ws_robot/meshes/ezmile_gen0_wheel.obj"/>
                </geometry>  
        </visual>
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${tyre_mass}"/>
            <inertia
            ixx="${tyre_mass_ixx}"  ixy="${tyre_mass_ixy}"  ixz="${tyre_mass_ixz}"
            iyy="${tyre_mass_iyy}"  iyz="${tyre_mass_iyz}"
            izz="${tyre_mass_izz}"/>
        </inertial>
    </link>

    <!--Left Steering-->
    <joint name="front_left_steering_joint" type="revolute">
        <parent link="base_link"/>
        <child link="front_left_steering_link"/>
        <origin xyz="${tyre_front_x} ${car_width/2 + 2*str_length} ${tyre_radius}" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="10000" lower="${-str_angle}" upper="${str_angle}" velocity="1000"/>
    </joint>

    <link name="front_left_steering_link">
        <parent link="base_link"/>
        <collision>
            <xacro:insert_block name="tyre_origin"/>
            <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
            </geometry>
        </collision>
        <visual>
            <xacro:insert_block name="tyre_origin"/>
            <geometry>
                <cylinder length="${str_length}" radius="${str_radius}"/>
            </geometry>
        </visual>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${str_mass}"/>
            <inertia
            ixx="${str_mass_ixx}"  ixy="0.000000"  ixz="0.000000"
            iyy="${str_mass_iyy}"  iyz="0.000000"
            izz="${str_mass_izz}"/>
        </inertial>
    </link>



<!--Front Left Wheel-->

    <joint name="front_left_wheel_joint" type="continuous">
          <parent link="front_left_steering_link"/>
          <child link="front_left_wheel_link"/>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <axis xyz="0 1 0"/>
          <dynamics damping="0.01"/>
          <limit effort="10000" velocity="5000" />
          <joint_properties damping="10" friction="1.0" />

    </joint>

    <link name="front_left_wheel_link">
        <parent link="front_left_wheel_joint"/>
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <!--origin xyz="0 0 0" rpy="0 0 ${PI/2}"/-->
            <geometry>
                <cylinder length="${tyre_length}" radius="${tyre_radius}"/>
                <!--mesh filename="package://ezmile_gen0/meshes/ford_escape_wheel.dae"/-->
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>10000000</mu>
                        <mu2>7000</mu2>
                    </ode>
                </friction>
                <bounce>
                    <restitution_coefficient>0.0</restitution_coefficient>
                </bounce>
                <contact>
                    <ode>
                        <soft_cfm>0.001</soft_cfm>
                        <soft_erp>0.5</soft_erp>
                        <kp>10000000</kp>
                        <kd>1</kd>
                    </ode>
                </contact>
            </surface>
        </collision>
        
        <visual>
            <!--origin xyz="0 0 0" rpy="${PI/2} 0 0"/-->
            <origin xyz="0 ${-car_width_offset*2} 0.0" rpy="0 0 ${PI/2}"/>
            <geometry>
            <!--cylinder length="${tyre_length}" radius="${tyre_radius}"/-->
            <mesh scale="1.2 1 1" filename="package://4ws_robot/meshes/ezmile_gen0_wheel.obj"/>
            </geometry>
        </visual>

        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${tyre_mass}"/>
            <inertia
            ixx="${tyre_mass_ixx}"  ixy="${tyre_mass_ixy}"  ixz="${tyre_mass_ixz}"
            iyy="${tyre_mass_iyy}"  iyz="${tyre_mass_iyz}"
            izz="${tyre_mass_izz}"/>
        </inertial>
    </link>

<!-- motors and transmissions for the two rear wheels -->
    <transmission name="tran1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="back_right_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor1">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    
    <transmission name="tran2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="back_left_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor2">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran3">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="back_left_steering_joint">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        
        <actuator name="eps_b_left">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

        <transmission name="tran4">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="back_right_steering_joint">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        
        <actuator name="eps_b_right">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    
    <transmission name="rack_and_pinion_left">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="front_left_steering_joint">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        
        <actuator name="eps_left">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="rack_and_pinion_right">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="front_right_steering_joint">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        
        <actuator name="eps_right">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
    <robotNamespace>/4ws</robotNamespace>
    <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
  </plugin>
</gazebo>

</robot>

